### DISCUSSION TITLE
Extension 컨벤션

### DISCUSSION THREAD

**Discussion by JIWON1923** [Team-ShortcutsZip/MacC-Team-HappyAnding#381] [General]
다른 프로젝트를 참고하며 얻은 인사이트 또는 개인의 의견을 공유해봅시다!

지난 회의(23.02.03)에서 나온 의견은 다음과 같습니다. (구분: 같은 파일로서 존재한다를 의미합니다)
 - **자료형을 기준**으로 구분하는 방법
 - **기능을 기준**으로 구분하는 방법
 - 자료형을 기준으로 구분하되, 기능을 기준으로 extension을 구분해서 작성하자

총 이렇게 3가지 의견이 존재했습니다!
다른 프로젝트들에서 확장을 어떻게 구분하고 있는지 조사하고, 더 직관적인 방법을 논의해봐요~!

**Comment by JIWON1923** [comment]
### 개인적인 의견
제가 extension 컨벤션을 제안했던 이유는, 자료형 기반의 확장이 어느순간부터 기능의 확장으로 작성되고 있다고 생각했기 때문입니다. 

저희 Date+Extension에 있는 `getVersionUpdateDateFormat()`을 기준으로 얘기해보겠습니다!
(이 함수는 "20230204"을 "2023.02.04"로 바꿔주는 함수입니다)

우선 `getVersionUpdateDateFormat()`가 수정되어야하는 경우는 너무 쉬웠습니다. ("2023-02-04"로 수정되는 등 함수 자체의 수정이 필요한 경우)
Xcode에서 기본으로 제공해주는 Jump to definition을 이용할 수도 있고, 다음과 같이 Quick help를 통해 접근할 수도 있었습니다.

![image](https://user-images.githubusercontent.com/68676844/216752740-48150eb7-26bf-4dba-83f3-a10962231e14.png)
---

하지만 `getVersionUpdateDateFormat()`를 사용해야하는 경우는 다르다고 생각했습니다.
"이런 함수가 이미 정의되어 있나?"를 확인할 때에는 Date extension 보다는 String extension을 찾아보는것이 자연스러운 플로우라고 생각합니다.

현재 가지고 있는 값이 String이기도 하고, 사용가능한 메서드 등을 표시해주는 창에도 String 관련된 것만 나타나기 때문입니다.
그래서 제 개인적인 의견은 **자료형을 기반으로 구분**하는 것이 맞다고 생각합니다!

<img src = "https://user-images.githubusercontent.com/68676844/216753074-27b3ff6d-f4fd-4f18-b7e5-75d555f47764.png" width = 500>

---

### 타 프로젝트 조사

**자료형을 기준**으로 구분하고 있었습니다. 
다만 저희처럼 하나의 Extension 파일에서 여러 기능을 제공하는 경우는 없었습니다. 
(String에서 "날짜"관련 정보를 다루기는 했지만, 날짜, 색상, 줄바꿈 등을 여러가지를 다루는 경우는 찾기 어려웠습니다.)
- [가슴 속 3천원](https://github.com/3dollar-in-my-pocket/3dollars-in-my-pocket-ios)
- [Firefighter 팀](https://github.com/DeveloperAcademy-POSTECH/MC2-Team5-Firefighter)
- [recipes 앱](https://github.com/mecid/swiftui-recipes-app)

---

### 결론
확장 자체는 자료형을 기반으로 파일을 구분해야한다고 생각합니다.
다만 타 프로젝트와 달리, 하나의 자료형에서 너무 많은 기능을 제공하고 있기 때문에 
extension을 구분해서 작성하며, **주석을 사용**하는 방법은 어떨까 제안합니다!
즉 **3번 의견 + 주석 달기**

---

### 추가 의견
- 제가 예시로 들었던 `getVersionUpdateDateFormat()`의 경우, ReadShortcutsView에서만 사용하고 있는데, Extension으로 정의하는 것이 적절할까요?
- 또한 사용되지 않는 extension 코드들이 꽤 존재하여 정리를 해야할 듯 싶습니다~!

**Reply by JIWON1923** [reply]
extension을 기능으로 구분하는 프로젝트들의 예시들도 찾아보겠습니다!!

**Comment by HanGyeongjun** [comment]
swiftui 태그가 붙은 프로젝트 중 star 상위권 프로젝트를 살펴보다 깔끔하게 폴더링이 되어있는 프로젝트를 찾아 공유합니다! 
https://github.com/CodeEditApp/CodeEdit

이 프로젝트도 extension을 데이터 타입으로 구분해 작성했는데, 자료형에 따라 폴더링을 하고 그 안에서 분리된 swift 파일로 extension을 작성하는 구조였습니다.
저희 프로젝트도 하나의 자료형에서 여러 기능을 제공하는 extension들이 많기 때문에 주석과 함께 이 방법도 검토해보면 괜찮을 것 같습니다!

자료형별 폴더링
<img width="1240" alt="Screenshot 2023-02-04 at 17 23 16" src="https://user-images.githubusercontent.com/94854258/216757149-1250b918-98ae-49ce-9f94-0b3a0178b4f1.png">

자료형 폴더 안에 각각의 swift 파일로 extension 작성
<img width="1236" alt="Screenshot 2023-02-04 at 17 23 37" src="https://user-images.githubusercontent.com/94854258/216757152-6c587721-e2e9-4dc5-8382-7a682ea90fc0.png">

**Reply by JIWON1923** [reply]
### 추가 정보
오호 이분들은 기능과 자료형을 모두 구분할 수 있도록 구현했군요!

조금 더 자료를 찾아보니, 로겐이 찾아주신 방식이 **Objective C에서 비롯된 방식**이라고 하네요!
Objc의 경우, 특정 클래스의 수평적 확장을 **Category라는 개념**을 통해서 진행한다고 해요.
이 카테고리의 이름을 + 뒤에 명시하게 됩니다 (필수적 네이밍 규칙은 아니지만, 권장했던 규칙이라고 생각하시면 될 것 같아요)

많은 분들이 [Stack overflow](https://stackoverflow.com/questions/26319660/whats-the-best-practice-for-naming-swift-files-that-add-extensions-to-existing)에서 많은 이야기를 나눴네요!!

가장 많은 공감을 얻은 의견은 로겐이 말씀해주신 것 처럼 **파일명을 보고 자료형과 목적을 한 번에 이해할 수 있는 구조**였어요

또 새로운 방법은 모든 String에서 사용하는 확장은 `String+extension`에 정의하고, 일반적이지 않은 경우에만 구분하여 선언해주는 방법도 있네요!
저희 프로젝트에서는 `String+Extension.swift`, `String+Date.swift` 이런식으로 구분할 수 있을 것 같아요!

만약 카테고리 명을 사용하게 된다면, 카테고리 명명 규칙이나 분류 기준에 대해서도 이야기 나눠봅시다~


### 참고한 사이트
[Objc 공식문서](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/CustomizingExistingClasses/CustomizingExistingClasses.html)
[Objc 카테고리](https://babbab2.tistory.com/73)
[Optimizing-Swift-Build-Times](https://github.com/fastred/Optimizing-Swift-Build-Times)
